url43a0865bb0_kevinjdolan_darwin_tgz-pJ8-tui_DriverJ8/src/darwin/work/threadPool/ThreadPool.java

```diff
@@ -0,0 +1,129 @@
+package darwin.work.threadPool;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.Vector;
+
+import darwin.work.Job;
+import darwin.work.JobRunner;
+import darwin.work.BusyException;
+
+/**
+ * The ThreadPool uses a finite number of threads to distribute
+ * the work
+ * @author 
+ */
+public class ThreadPool extends Thread implements JobRunner {
+
+	private Vector<Object> result;
+	private Queue< Pair<Job, Object> > queue;
+	private Worker[] workers;
+	private boolean running;
+	private HashMap<String, Object> shared;
+	private boolean alive;
+	
+	public ThreadPool(int numThreads) {
+		super("Pool");
+		queue = new ConcurrentLinkedQueue< Pair<Job, Object> >();
+		workers = new Worker[numThreads];
+		for(int i = 0; i < numThreads; i++)
+			workers[i] = new Worker(this);
+		running = false;
+		alive = true;
+		shared = new HashMap<String, Object>();
+		start();
+	}
+	
+	@Override
+	public void addJob(Job job, Object parameters) throws BusyException {
+		if(running)
+			throw new BusyException();
+		queue.add(new Pair<Job, Object>(job, parameters));
+	}
+
+	@Override
+	public void putSharedData(String name, Object value) {
+		shared.put(name, value);
+		for(Worker worker : workers)
+			worker.putSharedData(name, value);
+	}
+
+	@Override
+	public List<Object> dispatch() throws BusyException {
+		if(running)
+			throw new BusyException();
+		
+		running = true;
+		result = new Vector<Object>();
+		
+		//Wait for completion
+		while(running) {
+			try { sleep(100); }
+			catch(InterruptedException e) {}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)